File: lamps.py
"""
ID: totheso1
LANG: PYTHON3
TASK: lamps
"""

from itertools import combinations, cycle, islice
from typing import Any, Tuple, List

fin = open("lamps.in", "r")
fout = open("lamps.out", "w")


def fprint(*args: Any, **kwargs: Any) -> None:
    print(*args, file=fout, **kwargs)


# Code start
button_effect_collection = [0b111111, 0b101010, 0b010101, 0b100100]


def toggle(state: int, switch_collection: Tuple[int, ...]) -> int:
    for switch in switch_collection:
        state ^= button_effect_collection[switch]
    return state


MAXLAMPS = 6
MAXSWITCH = 4
N = int(fin.readline())
C = int(fin.readline())

bit_mask = 0
bit_requirement = 0
for i in map(int, fin.readline().split()):
    if i != -1:
        bit_mask |= 1 << (-i % MAXLAMPS)
        bit_requirement |= 1 << (-i % MAXLAMPS)
for i in map(int, fin.readline().split()):
    if i != -1:
        bit_mask |= 1 << (-i % MAXLAMPS)

ans: List[int] = []
for cnt in range(C & 1, min(MAXSWITCH + 1, C + 1), 2):
    for switch_collection in combinations(range(0, MAXSWITCH), cnt):
        state = toggle(button_effect_collection[0], switch_collection)
        if state & bit_mask == bit_requirement:
            ans.append(state)
if ans:
    ans.sort()
    for state in ans:
        fprint("".join(islice(cycle(f"{state:0>{MAXLAMPS}b}"), N)))
else:
    fprint("IMPOSSIBLE")
# Code end

File: lamps.cpp
/*
ID: totheso1
LANG: C++14
TASK: lamps
*/

#include <algorithm>
#include <bitset>
#include <fstream>
#include <iterator>
#include <vector>

using namespace std;

static ifstream fin("lamps.in");
static ofstream fout("lamps.out");

#define endl '\n'

template <typename T>
T fin_get() {
  return *istream_iterator<T>(fin);
}

template <typename C>
C fin_get_collection(const int size) {
  C ret;
  copy_n(istream_iterator<typename C::value_type>(fin), size,
         back_inserter(ret));
  return ret;
}

static constexpr int MAXLAMPS = 6;
static constexpr int MAXSWITCH = 4;

static int mod(const int dividend, const int divisor) {
  const int res = dividend % divisor;
  return res >= 0 ? res : res + divisor;
}

int main() {
  const int N = fin_get<int>();
  const int C = fin_get<int>();
  bitset<MAXLAMPS> bit_mask, bit_requirement;
  for (int i; (i = fin_get<int>()) != -1;) {
    bit_mask[mod(-i, MAXLAMPS)] = true;
    bit_requirement[mod(-i, MAXLAMPS)] = true;
  }
  for (int i; (i = fin_get<int>()) != -1;) {
    bit_mask[mod(-i, MAXLAMPS)] = true;
  }

  const vector<bitset<MAXLAMPS>> button_effect_collection = {
      0b111111, 0b101010, 0b010101, 0b100100};
  vector<bitset<MAXLAMPS>> ans;
  for (int cnt = C & 1, e = min(MAXSWITCH + 1, C + 1); cnt < e; cnt += 2) {
    vector<bool> switch_collection(MAXSWITCH, false);
    fill_n(begin(switch_collection), cnt, true);
    do {
      bitset<MAXLAMPS> state = button_effect_collection[0];
      int i = -1;
      for (const bool switch_on : switch_collection) {
        ++i;
        if (switch_on) {
          state ^= button_effect_collection[i];
        }
      }
      if ((state & bit_mask) == bit_requirement) {
        ans.push_back(move(state));
      }
    } while (
        prev_permutation(begin(switch_collection), end(switch_collection)));
  }
  if (ans.empty()) {
    fout << "IMPOSSIBLE" << endl;
  } else {
    sort(begin(ans), end(ans),
         [](const bitset<MAXLAMPS> &lhs, const bitset<MAXLAMPS> &rhs) {
           return lhs.to_ulong() < rhs.to_ulong();
         });
    for (const auto &state : ans) {
      for (int i = 1; i <= N; ++i) {
        fout << state[mod(-i, MAXLAMPS)];
      }
      fout << endl;
    }
  }
}
